### PR TITLE
Added search promotion button in omnibox

### PR DESCRIPTION
**Pull request by simonhong**
fix https://github.com/brave/brave-browser/issues/40776

image

Resolves

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Manual test: see the linked issue
`BraveLocationBarViewBrowserTest.SearchConversionButtonTest`

### CODE REVIEW

File: chromium_src/chrome/browser/ui/views/location_bar/location_bar_view.cc [hunk 1]
```diff
@@ -36,6 +36,15 @@
     add_trailing_decoration(item, /*intra_item_padding=*/0);    \
   }
 
+#define BRAVE_LAYOUT_HANDLE_SEARCH_PROMOTION_BUTTON_VISIBILITY              \
+  }                                                                         \
+  else if (GetSearchPromotionButton() &&                                    \
```
Comment (simonhong) [inline chromium_src/chrome/browser/ui/views/location_bar/location_bar_view.cc:41]: Can't avoid this presubmit error. @goodov can you advice how should I handle this?
```
brave/chromium_src/chrome/browser/ui/views/location_bar/location_bar_view.cc:41:  (cpplint) If an else has a brace on one side, it should have it on both  [readability/braces] [5]
```

File: browser/ui/views/location_bar/brave_search_conversion/promotion_button_controller.cc [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+          omnibox_view_->GetText());
+
+  NavigateParams params(&*browser_, url, ui::PAGE_TRANSITION_TYPED);
+  params.disposition = WindowOpenDisposition::CURRENT_TAB;
+  Navigate(&params);
+}
+
+void PromotionButtonController::Dismissed() {
+  button_->SetVisible(false);
+  brave_search_conversion::SetDismissed(&*prefs_);
```
Comment (fallaciousreasoning) [inline browser/ui/views/location_bar/brave_search_conversion/promotion_button_controller.cc:189]: Should this have a more specific name? I think we've got a few different conversion methods, which presumably use different prefs?

`SetDismissedLocationBarButton` or something?

**Review by fallaciousreasoning — APPROVED** [review]

File: browser/ui/views/location_bar/brave_search_conversion/promotion_button_controller.cc (hunk 2 above)
Comment (simonhong) [reply browser/ui/views/location_bar/brave_search_conversion/promotion_button_controller.cc:189]: This is intended as user dismissed already via another promotion, I think we don't want to show this button again.
@rebron WDYT?

File: browser/ui/color/brave_color_mixer.cc [hunk 3]
```diff
@@ -311,6 +311,26 @@ void AddBraveColorMixerForAllThemes(ui::ColorProvider* provider,
   // color by their color mixers.
   mixer[kColorToolbarButtonActivated] = {SkColorSetRGB(0x7C, 0x91, 0xFF)};
   mixer[kColorSidebarButtonPressed] = {kColorToolbarButtonActivated};
+
+  // Search conversion button in omnibox.
+  mixer[kColorSearchConversionButtonText] = {nala::kColorPrimary60};
```
[the remark this answers is not in the document]
Comment (simonhong) [reply browser/ui/color/brave_color_mixer.cc:316]: If we can adjust to our colors similar to custom theme, that would be great!

File: browser/ui/views/location_bar/brave_search_conversion/promotion_button_controller.cc (hunk 2 above)
Comment (fallaciousreasoning) [reply browser/ui/views/location_bar/brave_search_conversion/promotion_button_controller.cc:189]: Okay, that makes sense :)

File: chromium_src/chrome/browser/ui/views/location_bar/location_bar_view.cc (hunk 1 above)
Comment (goodov) [reply chromium_src/chrome/browser/ui/views/location_bar/location_bar_view.cc:41]: `// NOLINT` ?

File: patches/chrome-browser-ui-views-location_bar-location_bar_view.cc.patch [hunk 4]
```diff
@@ -10,15 +10,23 @@ index 61d32b558b7c49fa69989219b660bf2adf8fb90c..238b7d421119e21b15e95a86c789f5a2
    if (should_indent) {
      icon_left += icon_indent;
      text_left += text_indent;
-@@ -802,6 +803,7 @@ void LocationBarView::Layout(PassKey) {
+@@ -778,6 +779,7 @@ void LocationBarView::Layout(PassKey) {
+       }
+       selected_keyword_view_->SetCustomImage(image);
+     }
++  BRAVE_LAYOUT_HANDLE_SEARCH_PROMOTION_BUTTON_VISIBILITY
```
Comment (goodov) [inline patches/chrome-browser-ui-views-location_bar-location_bar_view.cc.patch:17]: nit: `BRAVE_LOCATION_BAR_VIEW_LAYOUT_...` (all defines in this file can be updated).

**Review by goodov — APPROVED** [review]

File: browser/ui/views/location_bar/brave_search_conversion/promotion_button_view.h [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+  void OnButtonPressed();
+  void OnClosePressed();
+
+  raw_ptr<views::ImageView> target_provider_image_ = nullptr;
+  std::unique_ptr<ViewShadow> shadow1_;
+  std::unique_ptr<ViewShadow> shadow2_;
+  std::unique_ptr<gfx::SlideAnimation> animation_;
+  base::OnceClosure dismissed_callback_;
+  base::OnceClosure make_default_callback_;
```
[the remark this answers is not in the document]
Comment (rebron) [reply browser/ui/views/location_bar/brave_search_conversion/promotion_button_view.h]: We shouldn't show the promotion button again if the user already set or dismissed it already.

**Pull request merged**